Adds Teradata translation

edgararuiz opened the pull request:
This change includes the new Teradata translation, and its corresponding tests.  

The new translations was tested using `dbtest`, here are the results:

![teradata-test](https://user-images.githubusercontent.com/7875923/31794609-7cef96d0-b4e8-11e7-91ba-ff6b8ff7b208.png)

Fixes: https://github.com/tidyverse/dplyr/issues/3040

hadley commented on R/translate-sql-odbc.R:
```diff
@@ -9,6 +9,7 @@ base_odbc_scalar <- sql_translator(.parent = base_scalar,
   as.character  = sql_cast("STRING"),
   as.Date       = sql_cast("DATE"),
   paste0        = sql_prefix("CONCAT"),
+  sd            = sql_prefix("STDDEV_SAMP"),
```
Are you sure this is ok to add here?

hadley commented on R/db-odbc-teradata.R:
```diff
@@ -0,0 +1,94 @@
+                        build_sql(
+                          "SUBSTR(", x, ", ", start, ", ", len, ")"
+                        )},
+      paste         = sql_not_supported("paste()")
```
https://forgetcode.com/Teradata/434-CONCAT ?

edgararuiz replied:
Right, I use `CONCAT` for `paste0`.  `CONCAT` does not add a separator like `paste` does by default.
https://github.com/tidyverse/dbplyr/blob/d08d93071606db414b44f8feb84dae44568c0300/R/translate-sql-odbc.R#L11

hadley replied:
Ah got it. Might be better to have a more helpful error here? i.e. try `paste0()` instead

edgararuiz replied:
Good idea, I'd like to create a function for that, can I add `sql_not_supported()` a new `instead` argument? Or should I create a new function?

hadley replied:
I think you can just create a new function inline

hadley commented on R/db-odbc-teradata.R:
```diff
@@ -0,0 +1,94 @@
+                        if (isTRUE(all.equal(base, exp(1)))) {
+                          build_sql("ln(", x, ")")
+                        } else {
+                          # Use log change-of-base because postgres doesn't support the
```
Is this true for teradata too? If so we should pull out into a resuable function

edgararuiz replied:
Ok, where should I place the function? I'm thinking `traslate-sql-helpers.R`

hadley replied:
Yeah, sounds good.

edgararuiz commented:
Ok, I moved `log` to its own function.  I noticed that `cot` was in the same situation so I moved that formula as well.  I also added, and updated,  the corresponding tests.

PTAL @hadley

hadley commented on R/translate-sql-helpers.r:
```diff
@@ -170,3 +170,22 @@ sql_cast <- function(type) {
+
+#' @rdname sql_variant
+#' @export
+sql_formula_log <- function(f, base = exp(1)) {
```
I think these can just be `sql_log` and `sql_cot`, and like `sql_cast()` etc should return functions

hadley commented on tests/testthat/test-translate-teradata.r:
```diff
@@ -0,0 +1,150 @@
+    sql("CAST(`field_name` AS VARCHAR(MAX))")
+  )
+})
+test_that("log10() translates to LOG ", {
```
Now that we have `sql_log()` used in multiple places, there's no need to test for each individual driver that uses it.

hadley commented on tests/testthat/test-translate-teradata.r:
```diff
@@ -0,0 +1,150 @@
+
+# Teradata base_agg conversions -----------------------------------------
+
+test_that("sd() translates to STDEV ", {
```
Can you please bundle these all into one test, writing a function as needed to reduce duplication?

edgararuiz commented:
PTAL @hadley
